### PR TITLE
Interpolation for arbitrary grids

### PR DESCRIPTION
**Pull request by jleinonen**
# Earth2Studio Pull Request

## Description

* Add interpolation between arbitrary 2D lat-lon grids as `earth2studio.utils.interp.LatLonInterpolation`. This is implemented as a `torch.nn.Module` because initializing the mapping between the grids can be a bit slow but after that the interpolation of the values is very fast.
* Move the interpolation function from the CorrDiff implementation to `earth2studio.utils.interp` for easier reuse in e.g. other CorrDiff implementations.
* Add test for `LatLonInterpolation`.

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/earth2studio/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/earth2studio/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/earth2studio/issues) is linked to this pull request.

### CODE REVIEW

File: earth2studio/utils/interp.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    The mapped values can be on arbitrary grid, but the output grid should be
+    contained within the input grid.
+
+    Initializing the interpolation object can be somewhat slow, but interpolation is
+    fast and runs on the GPU once initialized. Therefore, prefer to reuse the
+    interpolation object when possible.
+
+    Parameters
+    ----------
+    lat0 : torch.Tensor [H_in, W_in]
```
Comment (NickGeneva) [inline earth2studio/utils/interp.py]: Update param names

File: test/utils/test_interp.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
+import numpy as np
+import pytest
+import torch
+
+from earth2studio.utils.interp import LatLonInterpolation
+
+
+@pytest.mark.parametrize("device", ["cpu", "cuda:0"])
+@pytest.mark.parametrize("input_type", ["zeros", "random", "gradient"])
+def test_interpolation(device, input_type):
```
Comment (NickGeneva) [inline test/utils/test_interp.py:26]: Any chance you can implement at unit test here that checks the qualitative value of the output.

So have maybe you have a 2x2 reference in a [0,1] box with some fixed values at the corners. Then you have a few points at like [0.25,0.25] [0.25,0.5] [0.5.0.25] and [0.5 0.5].

And confirm the outputs are analytically what one would expect

File: earth2studio/utils/interp.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
+import numpy as np
+import torch
+from numpy.typing import ArrayLike
+from scipy.interpolate import LinearNDInterpolator
+from torch import Tensor, nn
+
+
+def latlon_interpolation_regular(
```
Comment (NickGeneva) [inline earth2studio/utils/interp.py:24]: Confirming this is just a 1:1 copy from corrdiff yeah?

File: earth2studio/utils/interp.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+        i_map = i_interp(out_points).reshape(lat_out.shape)
+        j_map = j_interp(out_points).reshape(lat_out.shape)
+
+        i_map = torch.Tensor(i_map.clip(min=0, max=lat_in.shape[0] - 2))
+        j_map = torch.Tensor(j_map.clip(min=0, max=lat_in.shape[1] - 2))
+
+        self.register_buffer("i_map", i_map)
+        self.register_buffer("j_map", j_map)
+
+    @torch.inference_mode()
```
Comment (NickGeneva) [inline earth2studio/utils/interp.py:164]: Can you try adding @torch.compile here

File: earth2studio/utils/interp.py [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+    def forward(self, values: Tensor) -> Tensor:
+        """Perform bilinear interpolation for values.
+
+        Parameters
+        ----------
+        values : torch.Tensor [..., H_in, W_in]
+            Input values defined over (lat0, lon0) that will be interpolated onto
+            (lat1, lon1) grid.
```
Comment (NickGeneva) [inline earth2studio/utils/interp.py]: Can you add a return doc here so it tells the expected dim
```
Returns
-------
torch.Tensor
    Interpolated tensor of shape [..., W_out, H_out]
```

File: earth2studio/utils/interp.py [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+        self.register_buffer("i_map", i_map)
+        self.register_buffer("j_map", j_map)
+
+    @torch.inference_mode()
+    def forward(self, values: Tensor) -> Tensor:
+        """Perform bilinear interpolation for values.
+
+        Parameters
+        ----------
+        values : torch.Tensor [..., H_in, W_in]
```
Comment (NickGeneva) [inline earth2studio/utils/interp.py]: Nit:

Move the dims ([..., H_in, W_in]) into the description line

File: earth2studio/utils/interp.py [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+    fast and runs on the GPU once initialized. Therefore, prefer to reuse the
+    interpolation object when possible.
+
+    Parameters
+    ----------
+    lat0 : torch.Tensor [H_in, W_in]
+        Tensor of input latitude coordinates
+    lon0 : torch.Tensor [H_in, W_in]
+        Tensor of input longitude coordinates
+    lat1 : torch.Tensor [W_out, H_out]
```
Comment (NickGeneva) [inline earth2studio/utils/interp.py]: Same Nit as below, dims down into description. And update to torch.Tensor | ArrayLike

File: earth2studio/utils/interp.py [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+        Tensor of output latitude coordinates
+    lon1 : torch.Tensor [W_out, H_out]
+        Tensor of output longitude coordinates
+    """
+
+    def __init__(
+        self,
+        lat_in: Tensor | ArrayLike,
+        lon_in: Tensor | ArrayLike,
```
Comment (NickGeneva) [inline earth2studio/utils/interp.py]: Nit: typically like to use torch.Tensor here just so its explicit what tensor/package its expecting

**Comment by NickGeneva** [comment]
Can you also add this to the API docs: https://github.com/NVIDIA/earth2studio/blob/main/docs/modules/utils.rst#modearth2studioutils-coordinates

Under the coordinate section under the utils, can add:

```
.. autosummary::
   :toctree: generated/utils/
   :template: class.rst

 utils.interp.LatLonInterpolation
```

You can make just the API docs with:
```
pip install -e .[docs]
make docs
cd docs/_build/html
```

**Comment by NickGeneva** [comment]
Lastly add a note to the change log that this was added!

**Comment by NickGeneva** [comment]
Just some minor nits / edits and some doc items. Looks great to me otherwise, thanks Jussi

File: earth2studio/utils/interp.py (hunk 3 above)
Comment (jleinonen) [reply earth2studio/utils/interp.py:24]: Yes, it's just cut/pasted over from corrdiff.

File: earth2studio/utils/interp.py (hunk 8 above)
Comment (jleinonen) [reply earth2studio/utils/interp.py]: Changed in https://github.com/NVIDIA/earth2studio/pull/159/commits/cc59471fae79f5305c9e0cb512ca4d56e0f76834.

File: earth2studio/utils/interp.py (hunk 5 above)
Comment (jleinonen) [reply earth2studio/utils/interp.py]: Changed in https://github.com/NVIDIA/earth2studio/pull/159/commits/cc59471fae79f5305c9e0cb512ca4d56e0f76834.

File: earth2studio/utils/interp.py (hunk 1 above)
Comment (jleinonen) [reply earth2studio/utils/interp.py]: Changed in https://github.com/NVIDIA/earth2studio/pull/159/commits/cc59471fae79f5305c9e0cb512ca4d56e0f76834.

File: earth2studio/utils/interp.py (hunk 6 above)
Comment (jleinonen) [reply earth2studio/utils/interp.py]: Changed in https://github.com/NVIDIA/earth2studio/pull/159/commits/4df40a1f4ca4dba0c8da60f9d9fc6a6544546754.

File: earth2studio/utils/interp.py (hunk 7 above)
Comment (jleinonen) [reply earth2studio/utils/interp.py]: Changed in https://github.com/NVIDIA/earth2studio/pull/159/commits/4df40a1f4ca4dba0c8da60f9d9fc6a6544546754.

File: earth2studio/utils/interp.py (hunk 4 above)
Comment (jleinonen) [reply earth2studio/utils/interp.py:164]: Added in https://github.com/NVIDIA/earth2studio/pull/159/commits/4df40a1f4ca4dba0c8da60f9d9fc6a6544546754, seems to work.

File: test/utils/test_interp.py (hunk 2 above)
Comment (jleinonen) [reply test/utils/test_interp.py:26]: Added in https://github.com/NVIDIA/earth2studio/pull/159/commits/4df40a1f4ca4dba0c8da60f9d9fc6a6544546754.

**Comment by jleinonen** [comment]
> Can you also add this to the API docs: https://github.com/NVIDIA/earth2studio/blob/main/docs/modules/utils.rst#modearth2studioutils-coordinates
> 
> Under the coordinate section under the utils, can add:
> 
> ```
> .. autosummary::
>    :toctree: generated/utils/
>    :template: class.rst
> 
>  utils.interp.LatLonInterpolation
> ```
> 
> You can make just the API docs with:
> 
> ```
> pip install -e .[docs]
> make docs
> cd docs/_build/html
> ```

Added in https://github.com/NVIDIA/earth2studio/pull/159/commits/e0a7395fe498a3225d269a2b0aaa8c6f0d8ac445, could you check that it's correct?

> Lastly add a note to the change log that this was added!

Also added in the above commit.

File: test/utils/test_interp.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
+import numpy as np
```
Comment (NickGeneva) [inline test/utils/test_interp.py:17]: Ooops this file needs a license header. Thanks!

**Review by NickGeneva — APPROVED** [review]
